// @microsoft/teamsfx-core/build/component/generator/officeXMLAddin/projectConfig.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.OfficeAddinProjectConfig = void 0;
exports.OfficeAddinProjectConfig = {
    json: {
        "json-taskpane": {
            title: "core.newTaskpaneAddin.label",
            detail: "core.newTaskpaneAddin.detail",
            localTemplate: "",
            framework: {
                default_old: {
                    typescript: "https://aka.ms/teams-toolkit/office-addin-taskpane",
                },
                default: {
                    typescript: "https://aka.ms/teams-toolkit/office-addin-taskpane/ts-default",
                    javascript: "https://aka.ms/teams-toolkit/office-addin-taskpane/js-default",
                },
                react: {
                    typescript: "https://aka.ms/teams-toolkit/office-addin-taskpane/ts-react",
                    javascript: "https://aka.ms/teams-toolkit/office-addin-taskpane/js-react",
                },
            },
            manifestPath: "manifest.json",
        },
        "office-content-addin": {
            title: "core.newContentAddin.label",
            detail: "core.newContentAddin.detail",
            localTemplate: "",
            framework: {
                default: {
                    typescript: "https://aka.ms/teams-toolkit/office-addin-content/ts-default",
                    javascript: "https://aka.ms/teams-toolkit/office-addin-content/js-default",
                },
            },
            manifestPath: "manifest.json",
        },
    },
};
