// office-addin-lint/lib/defaults.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.usageDataObject = exports.PrettierExitCode = exports.ESLintExitCode = exports.lintFiles = void 0;
const office_addin_usage_data_1 = require("office-addin-usage-data");
exports.lintFiles = "src/**/*.{ts,tsx,js,jsx}";
var ESLintExitCode;
(function (ESLintExitCode) {
    ESLintExitCode[ESLintExitCode["NoLintErrors"] = 0] = "NoLintErrors";
    ESLintExitCode[ESLintExitCode["HasLintError"] = 1] = "HasLintError";
    ESLintExitCode[ESLintExitCode["CommandFailed"] = 2] = "CommandFailed";
})(ESLintExitCode || (exports.ESLintExitCode = ESLintExitCode = {}));
var PrettierExitCode;
(function (PrettierExitCode) {
    PrettierExitCode[PrettierExitCode["NoFormattingProblems"] = 0] = "NoFormattingProblems";
    PrettierExitCode[PrettierExitCode["HasFormattingProblem"] = 1] = "HasFormattingProblem";
    PrettierExitCode[PrettierExitCode["CommandFailed"] = 2] = "CommandFailed";
})(PrettierExitCode || (exports.PrettierExitCode = PrettierExitCode = {}));
// Usage data defaults
exports.usageDataObject = new office_addin_usage_data_1.OfficeAddinUsageData({
    projectName: "office-addin-lint",
    instrumentationKey: office_addin_usage_data_1.instrumentationKeyForOfficeAddinCLITools,
    raisePrompt: false,
});
