// @microsoft/teamsfx-core/build/component/driver/teamsApp/interfaces/OauthRegistration.js
"use strict";
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
Object.defineProperty(exports, "__esModule", { value: true });
exports.OauthRegistrationUserAccessType = exports.OauthRegistrationTargetAudience = exports.OauthRegistrationAppType = void 0;
var OauthRegistrationAppType;
(function (OauthRegistrationAppType) {
    OauthRegistrationAppType["SpecificApp"] = "SpecificApp";
    OauthRegistrationAppType["AnyApp"] = "AnyApp";
})(OauthRegistrationAppType = exports.OauthRegistrationAppType || (exports.OauthRegistrationAppType = {}));
var OauthRegistrationTargetAudience;
(function (OauthRegistrationTargetAudience) {
    OauthRegistrationTargetAudience["HomeTenant"] = "HomeTenant";
    OauthRegistrationTargetAudience["AnyTenant"] = "AnyTenant";
})(OauthRegistrationTargetAudience = exports.OauthRegistrationTargetAudience || (exports.OauthRegistrationTargetAudience = {}));
var OauthRegistrationUserAccessType;
(function (OauthRegistrationUserAccessType) {
    OauthRegistrationUserAccessType["Read"] = "Read";
    OauthRegistrationUserAccessType["ReadWrite"] = "ReadWrite";
})(OauthRegistrationUserAccessType = exports.OauthRegistrationUserAccessType || (exports.OauthRegistrationUserAccessType = {}));
